src/executor/current_thread.rs

```diff
@@ -0,0 +1,361 @@
+    /// Return a new `TaskRunner`
+    fn new(thread_notify: Arc<ThreadNotify>) -> TaskRunner {
+        TaskRunner {
+            non_daemons: 0,
+            scheduler: scheduler::Scheduler::new(thread_notify),
+        }
+    }
+
+    /// Enter a new `TaskRunner` context
+    fn enter<F, A>(f: F) -> Result<A::Item, A::Error>
+        where F: FnOnce() -> A,
+              A: Future,
+    {
+        // Create a new task runner that will be used for the duration of `f`.
+        ThreadNotify::with_current(|thread_notify| {
+            // The runner has to be created outside of the MY_TASK_RUNNER.with
+            // block.
+            let mut runner = TaskRunner::new(thread_notify.clone());
+
+            CURRENT.with(|current| {
+                // Make sure that another task runner is not set.
+                assert!(current.scheduler.get().is_null());
+
+                // Set the scheduler to the TLS and perform setup work,
+                // returning a future to execute.
+                //
+                // This could possibly spawn other tasks.
+                let future = current.set_scheduler(&mut runner.scheduler, f);
+
+                // Execute the runner
+                runner.finish(thread_notify, current, future)
+            })
+        })
+    }
+
+    fn finish<F: Future>(&mut self,
+                         thread_notify: &Arc<ThreadNotify>,
+                         current: &CurrentRunner,
+                         future: F)
+        -> Result<F::Item, F::Error>
+    {
+        let mut result = None;
+        let mut future = Some(executor::spawn(future));
+
+        while future.is_some() || self.non_daemons > 0 {
+            if current.cancel.get() {
+                // TODO: This probably can be improved
+                current.cancel.set(false);
+
+                debug_assert!(current.scheduler.get().is_null());
+                self.scheduler = scheduler::Scheduler::new(thread_notify.clone());
+            }
+
+            let res = future.as_mut()
+                .map(|f| f.poll_future_notify(thread_notify, 0));
+
+            match res {
+                Some(Ok(Async::Ready(e))) => {
+                    result = Some(Ok(e));
+                    future = None;
+                }
+                Some(Err(e)) => {
+                    result = Some(Err(e));
+                    future = None;
+                }
+                Some(Ok(Async::NotReady)) |
+                None => {}
+            }
+
+            self.poll_all(current);
+
+            if future.is_some() || self.non_daemons > 0 {
+                thread_notify.park();
+            }
+        }
+
+        result.unwrap()
+    }
+
+    fn poll_all(&mut self, current: &CurrentRunner) {
+        use scheduler::Tick;
+
+        loop {
+            let res = self.scheduler.tick(|scheduler, spawned, notify| {
+                current.set_scheduler(scheduler, || {
+                    match spawned.inner.0.poll_future_notify(notify, 0) {
```

Comment:
I'd love some commentary here about the fact that `tick` gives us `&mut` access back to the scheduler, which provides some amount of safety checking against re-entrancy issues.